tests/unit/s2n_key_material_test.c

```diff
@@ -0,0 +1,209 @@
+    test_ptr += iv_size;
+    RESULT_ENSURE_REF(test_ptr);
+    /* server IV */
+    RESULT_ENSURE_EQ(memcmp(test_ptr, key_material->server_iv.data, iv_size), 0);
+
+    return S2N_RESULT_OK;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* prepare test data */
+    uint8_t test_data[S2N_MAX_KEY_BLOCK_LEN] = { 0 };
+    struct s2n_blob test_data_blob = { 0 };
+    EXPECT_SUCCESS(s2n_blob_init(&test_data_blob, test_data, sizeof(test_data)));
+    EXPECT_OK(helper_generate_test_data(&test_data_blob));
+
+    /* fuzz s2n_key_material_init with different mac, key, iv sizes */
+    {
+        for (uint8_t mac_size = 0; mac_size < 16; mac_size++) {
+            for (uint8_t key_size = 0; key_size < 32; key_size++) {
+                for (uint8_t iv_size = 0; iv_size < 32; iv_size++) {
+                    EXPECT_TRUE((mac_size * 2 + key_size * 2 + iv_size * 2 <= S2N_MAX_KEY_BLOCK_LEN));
+
+                    DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT),
+                            s2n_connection_ptr_free);
+
+                    /* test varying size of mac, key and iv */
+                    conn->actual_protocol_version = S2N_TLS10;
+                    struct s2n_cipher temp_cipher = {
+                        .type = S2N_COMPOSITE,
+                        .key_material_size = key_size,
+                        .io.comp = {
+                                /* interpreted as iv size for composite ciphers.. which makes
+                                 * it easy for testing */
+                                .block_size = iv_size,
+                                .mac_key_size = mac_size,
+                        },
+                    };
+                    struct s2n_record_algorithm temp_record_alg = {
+                        .cipher = &temp_cipher,
+                    };
+                    struct s2n_cipher_suite temp_cipher_suite = {
+                        .record_alg = &temp_record_alg,
+                    };
+                    conn->secure->cipher_suite = &temp_cipher_suite;
+
+                    /* init s2n_key_material */
+                    struct s2n_key_material key_material = { 0 };
+                    EXPECT_OK(s2n_key_material_init(&key_material, conn));
+
+                    /* assert that sizes match */
+                    EXPECT_EQUAL(key_material.client_mac.size, mac_size);
+                    EXPECT_EQUAL(key_material.client_key.size, key_size);
+                    EXPECT_EQUAL(key_material.client_iv.size, iv_size);
+                    EXPECT_EQUAL(key_material.server_mac.size, mac_size);
+                    EXPECT_EQUAL(key_material.server_key.size, key_size);
+                    EXPECT_EQUAL(key_material.server_iv.size, iv_size);
+
+                    /* copy data into key_material and validate accessing key_material is sound */
+                    POSIX_CHECKED_MEMCPY(key_material.key_block, test_data, sizeof(key_material.key_block));
```

Comment:
Careful with the size. Can you assert that sizeof(key_material.key_block) == sizeof(test_data) before you copy? Whenever you do a copy, you have to worry about memory safety.